### PR TITLE
Improve efficiency of some constant time functions

### PR DESCRIPTION
**Pull request by daverodgman**
Follow-up to #6666 - improve efficiency of our constant time memcmp and memcpy implementations.

## Gatekeeper checklist

- [x] **changelog** not required (covered by changelog entry from #6666)
- [x] **backport** not required
- [x] **tests** added tests for these functions

### CODE REVIEW

**Comment by tom-cosgrove-arm** [comment]
Bah, this has conflicts

**Comment by daverodgman** [comment]
Unfortunately this will have a small conflict with #6876, so will need another rebase when that gets merged.

**Comment by tom-cosgrove-arm** [comment]
#6876 has been merged, but I don't see a conflict reported

File: include/mbedtls/build_info.h
Comment (tom-cosgrove-arm) [inline include/mbedtls/build_info.h:67]: why do we need the `*INDENT-OFF*` and `*INDENT-ON*`?

Comment (gilles-peskine-arm) [reply include/mbedtls/build_info.h:67]: Because that particular construct confuses uncrustify's parser. Cf changelog message that introduced it originally.

File: include/mbedtls/build_info.h
Comment (gilles-peskine-arm) [inline include/mbedtls/build_info.h:64]: I don't think it's acceptable to mess with the definition of `asm` as a macro in a public header. This might conflict with other tricks that the application is pulling with `asm`.

Doing this in `library/common.h` would be acceptable.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
We're overstepping with the public definition of `asm`.

I have not reviewed this otherwise.

File: include/mbedtls/build_info.h
Comment (tom-cosgrove-arm) [reply include/mbedtls/build_info.h]: Wow, that's a pretty bad parser, as this is not an uncommon construct :( - and also, didn't appear to be necessary in `library/bn_mul.h` (i.e. it wasn't done there)

File: library/constant_time.c
Comment (tom-cosgrove-arm) [inline library/constant_time.c:90]: I know it was here previously, but I don't think `volatile` is necessary or helpful here.

Looking at compiler output, having `volatile` causes an extra store to memory in the stack frame; not having it doesn't seem to cause any warnings

Comment (daverodgman) [reply library/constant_time.c:90]: Looks like @gabor-mezei-arm added the `volatile` in db9a38c672 - do you know the reason for this Gabor?

Comment (gilles-peskine-arm) [reply library/constant_time.c:90]: Gábor was moving some preexisting code (formerly in `include/mbedtls/ssl_internal.h`). The `volatile` keyword for `diff` was added by Hanno in 59e6963a37a615b137ee4f9824f798dc704fdd96 via https://github.com/Mbed-TLS/mbedtls-restricted/pull/370.

Comment (tom-cosgrove-arm) [reply library/constant_time.c:90]: There are comments in that commit message that should really be comments in the code!

Comment (gabor-mezei-arm) [reply library/constant_time.c:90]: I moved the different  `memcmp` implementations and I kept the 'volatile' version from them. I don't remember the reason.

Comment (daverodgman) [reply library/constant_time.c:90]: Interesting, it wasn't in every implementation (there were a few impls of `memcmp` in various files which were unified in `constant_time.c`). E.g., `cipher.c` and `rsa.c` did not have it, but `nist_kw.c` did.

Comment (daverodgman) [reply library/constant_time.c:90]: I would prefer to remove it, unless anyone is able to suggest a reason that it might be necessary?

File: library/constant_time.c
Comment (daverodgman) [reply library/constant_time.c:77]: I'm not familiar with x86 asm so won't attempt to cover that in this PR. x86 perf will be unaffected if this isn't done.

File: library/constant_time.c
Comment (gilles-peskine-arm) [reply library/constant_time.c:90]: The divergence is because we had multiple copies and they didn't get updated at the same time (which is part of why we eventually unified them). When `volatile` was added to the SSL function in https://github.com/Mbed-TLS/mbedtls-restricted/pull/370, there was no other constant time memcmp. There was another copy in https://github.com/Mbed-TLS/mbedtls/pull/485 but it wasn't merged yet (that eventually happened via https://github.com/Mbed-TLS/mbedtls/pull/1617).

I am against removing `volatile` unless you find a flaw in the reasoning that led to adding it.

Comment (daverodgman) [reply library/constant_time.c:90]: From a quick read of #370: "theoretically opening the possibility for the compiler to notice that the loop operation diff |= A[i] ^ B[i] is pointless if diff = -1 (which happens very quickly if A and B are random and independent!)" - how would the compiler be able to determine this, given that A and B are volatile and therefore the compiler can't know their values at compile time?

Comment (gilles-peskine-arm) [reply library/constant_time.c:90]: > A and B are volatile and therefore the compiler can't know their values at compile time

This does not follow. `volatile` mandates that the read takes place, but not that the compiler can't make assumptions on the value that is read. For that you need to manage to pass the pointer to something that the compiler can't analyze.

Compilers generally don't try to optimize when `^` is involved (because it doesn't appear in “normal” code, so compiler writers don't bother). So this code is likely safe in practice even without `volatile`. But in principle, they can.

Would `volatile register` help to avoid a memory access here? I'm not sure modern compilers pay any attention to `register`.

Comment (daverodgman) [reply library/constant_time.c:90]: A and B are pointers-to-volatile, so the compiler can't optimise those reads away. So AIUI the compiler will always have to execute the full loop.

Comment (tom-cosgrove-arm) [reply library/constant_time.c:90]: If we keep `volatile` in I would really like us to add the reasoning to the code - it would have saved all of this discussion had it been there

Comment (tom-cosgrove-arm) [inline library/constant_time.c:90]: So how about something like

```
    /* `A` and `B` are volatile because otherwise, once `diff` gets to -1
     * (which happens ra[pidly if A and B are random and independent),
     * no further reading of either is necessary. */
    volatile const unsigned char *A = (volatile const unsigned char *) a;
    volatile const unsigned char *B = (volatile const unsigned char *) b;

    /* `diff` is volatile to ensure that the compiler always does the xor
    // (as it has to read and write it every iteration). */
    volatile unsigned char diff = 0;
```

Comment (tom-cosgrove-arm) [reply library/constant_time.c:90]: OpenSSL doesn't have the `volatile` on its equivalent of `diff`: https://github.com/openssl/openssl/blob/master/crypto/cpuid.c#L184

Comment (daverodgman) [reply library/constant_time.c:90]: I discussed with Gilles & we agreed that there's no meaningful risk from dropping the volatile.

Comment (daverodgman) [reply library/constant_time.c:90]: If there is no further objection I will drop `volatile` from the definition of `diff`.

Comment (tom-cosgrove-arm) [reply library/constant_time.c:90]: Yes please. And while I think it's obvious, should we comment on the `volatile` for `A` and `B`?

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM

File: library/constant_time.c
Comment (gilles-peskine-arm) [reply library/constant_time.c]: Consider the old code (because there's a lot less to follow), with a non-volatile `diff`:

```
    unsigned char diff = 0;
    for( i = 0; i < n; i++ ) {
        unsigned char x = A[i], y = B[i];
        diff |= x ^ y;
    }
    return( (int)diff );
```

A compiler is allowed to compile this as:
* while `i < n` and `x = A[i]; y = B[i]; diff != 0xff`: `diff |= A[i] ^ B[i]`
* while `i < n`: `x = A[i]; y = B[i];`
* return `diff`

Making `diff` volatile avoids this particular “optimization”.

I doubt that any existing compiler would look for such an “optimization” (which wouldn't make the code faster or smaller), but it would be allowed.

**Comment by daverodgman** [comment]
Force-push for code-style

File: library/alignment.h
Comment (tom-cosgrove-arm) [inline library/alignment.h:246]: I don't remember all this whitespace change in `alignment.h` - what's happened?

**Comment by tom-cosgrove-arm** [comment]
CI fail: `Failures: all_u18-check_code_style`

File: library/alignment.h
Comment (gilles-peskine-arm) [reply library/alignment.h:246]: The coding style change happened. Looks like it didn't go right? @daverodgman can you please share exactly what command you ran and the `--verbose` output?

Comment (daverodgman) [reply library/alignment.h]: Wrong uncrustify version I think

File: tests/scripts/all.sh
Comment (tom-cosgrove-arm) [inline tests/scripts/all.sh:3744]: Not sure why GitHub thinks this is part of the PR - probably not rebased off latest development?

Anyway, these changes are already in `development`, so should merge cleanly

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM - same as last time I approved

**Comment by tom-cosgrove-arm** [comment]
Hmm, this did have a passing CI, and no changes since then...

**Comment by daverodgman** [comment]
Rebase to get an up-to-date CI run and remove the commit from development that mysteriously made its way into this PR. Rebase was clean.

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM - confirmed that it's the same but with the redundant code that's already in `development` no longer present

**Comment by tom-cosgrove-arm** [comment]
Oops, still needs a second review (from an as-yet-unknown second reviewer)

File: library/constant_time.c
Comment (yanesca) [inline library/constant_time.c:77]: What is the difference between this and using `ldrh`?

**Review by yanesca — CHANGES_REQUESTED** [review]
Left a couple of remarks and a question inline.

File: library/constant_time.c
Comment (daverodgman) [reply library/constant_time.c:77]: `LDRH` would load 16 bits

File: tests/suites/test_suite_constant_time.function
Comment (tom-cosgrove-arm) [inline tests/suites/test_suite_constant_time.function:38]: So it's not really (boolean) `same` but `failing_offset` now?

File: tests/suites/test_suite_constant_time.function
Comment (tom-cosgrove-arm) [inline tests/suites/test_suite_constant_time.function:56]: What does `same >= size` mean? Isn't that "out of range"?

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM, although a couple of comments

File: tests/suites/test_suite_constant_time.function
Comment (daverodgman) [reply tests/suites/test_suite_constant_time.function:38]: Yes. I guess renaming it would be nicer, but I'll hold off unless there is more to do.

File: tests/suites/test_suite_constant_time.function
Comment (daverodgman) [reply tests/suites/test_suite_constant_time.function:56]: The first different byte is beyond the specified size, so in effect, same behaviour as `same == -1`. This change could have been omitted I suppose.

**Comment by tom-cosgrove-arm** [comment]
Note: full CI pass (all green!)

File: library/constant_time.c
Comment (yanesca) [reply library/constant_time.c:77]: Oh, so a word is 32 bits on aarch64 as well? Didn't know that.

**Review by yanesca — APPROVED** [review]
Left a minor remark, but other than that it looks good to me.

**Pull request merged**
